src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs

```diff
@@ -275,108 +275,174 @@ public static void ValidateHeaderNameCharacters(string headerCharacters)
             }
         }
 
-        public static unsafe ConnectionOptions ParseConnection(StringValues connection)
+        public static ConnectionOptions ParseConnection(StringValues connection)
         {
+            // Keep-alive
+            const ulong lowerCaseKeep = 0x0000_0020_0020_0020; // Don't lowercase hyphen
+            const ulong keepAliveStart = 0x002d_0070_0065_0065; // 4 chars "eep-"
+            const ulong keepAliveMiddle = 0x0076_0069_006c_0061; // 4 chars "aliv"
+            const ushort keppAliveEnd = 0x0065; // 1 char "e"
+            // Upgrade
+            const ulong upgradeStart = 0x0061_0072_0067_0070; // 4 chars "pgra"
+            const uint upgradeEnd = 0x0065_0064; // 2 chars "de"
+            // Close
+            const ulong closeEnd = 0x0065_0073_006f_006c; // 4 chars "lose"
+
             var connectionOptions = ConnectionOptions.None;
 
             var connectionCount = connection.Count;
             for (var i = 0; i < connectionCount; i++)
             {
-                var value = connection[i];
-                fixed (char* ptr = value)
+                var value = connection[i].AsSpan();
+                while (value.Length > 0)
                 {
-                    var ch = ptr;
-                    var tokenEnd = ch;
-                    var end = ch + value.Length;
-
-                    while (ch < end)
+                    int offset;
+                    char c = '\0';
+                    // Skip any spaces and empty values.
+                    for (offset = 0; offset < value.Length; offset++)
                     {
-                        while (tokenEnd < end && *tokenEnd != ',')
+                        c = value[offset];
+                        if (c == ' ' || c == ',')
                         {
-                            tokenEnd++;
+                            continue;
```

Comment:
Nit: Invert the if and break.

Reply:
Done

Reply:
Sweet! Do you mind updating the transfer-encoding PR to match? Thanks.

Reply:
Done